tonic-types/src/lib.rs

```diff
@@ -13,8 +23,391 @@
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             match status.check_error_details_vec() {
+    ///                 Ok(err_details) => {
+    ///                     // Handle extracted details
+    ///                 }
+    ///                 Err(decode_error) => {
+    ///                     // Handle decode_error
+    ///                 }
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn check_error_details_vec(&self) -> Result<Vec<ErrorDetail>, DecodeError>;
+
+    /// Get a vector of [`ErrorDetail`] enums from `tonic::Status`. If some
+    /// `prost::DecodeError` occurs, an empty vector will be returned.
+    /// # Examples
+    ///
+    /// ```
+    /// use tonic::{Status, Response};
+    /// use tonic_types::{ErrorDetail, WithErrorDetails};
+    ///
+    /// fn handle_request_result<T>(req_result: Result<Response<T>, Status>) {
+    ///     match req_result {
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             let err_details = status.get_error_details_vec();
+    ///             for err_detail in err_details.iter() {
+    ///                  match err_detail {
+    ///                     ErrorDetail::BadRequest(bad_request) => {
+    ///                         // Handle bad_request details
+    ///                     }
+    ///                     // ...
+    ///                     _ => {}
+    ///                  }
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn get_error_details_vec(&self) -> Vec<ErrorDetail>;
+
+    /// Get first [`BadRequest`] details found on `tonic::Status`, if any. If
+    /// some `prost::DecodeError` occurs, returns `None`.
+    /// # Examples
+    ///
+    /// ```
+    /// use tonic::{Status, Response};
+    /// use tonic_types::{WithErrorDetails};
+    ///
+    /// fn handle_request_result<T>(req_result: Result<Response<T>, Status>) {
+    ///     match req_result {
+    ///         Ok(_) => {},
+    ///         Err(status) => {
+    ///             if let Some(bad_request) = status.get_details_bad_request() {
+    ///                 // Handle bad_request details
+    ///             }
+    ///         }
+    ///     };
+    /// }
+    /// ```
+    fn get_details_bad_request(&self) -> Option<BadRequest>;
+}
+
+impl WithErrorDetails for tonic::Status {
+    fn with_error_details(code: Code, message: impl Into<String>, details: ErrorDetails) -> Self {
+        let message: String = message.into();
+
+        let mut conv_details: Vec<Any> = Vec::with_capacity(10);
+
+        if let Some(bad_request) = details.bad_request {
+            conv_details.push(bad_request.into_any());
+        }
+
+        let status = pb::Status {
+            code: code as i32,
+            message: message.clone(),
+            details: conv_details,
+        };
+
+        tonic::Status::with_details(code, message, Bytes::from(status.encode_to_vec()))
```

Comment:
This will overwrite any details from the previous status right? Do we want to do anything smart here? Like can we check if it was already a status message decode and add on to it?

Reply:
I'm not sure I understood you correctly, this associated function returns a new `Status` but we could implement a new method similar to it with this "add on" functionality

Reply:
Ah I mis-read this function, this seems fine to me.